mini/einstein_test.go

```diff
@@ -0,0 +1,194 @@
+package mini
+
+import (
+	"testing"
+	"time"
+
+	"github.com/awalterschulze/gominikanren/micro"
+	"github.com/awalterschulze/gominikanren/sexpr/ast"
+)
+
+/*
+Let us assume that there are five houses of different colors next to each other on the same road.
+In each house lives a man of a different nationality.
+Every man has his favorite drink, his favorite brand of cigarettes, and keeps pets of a particular kind.
+
+The Englishman lives in the red house.
+The Swede keeps dogs.
+The Dane drinks tea.
+The green house is just to the left of the white one.
+The owner of the green house drinks coffee.
+The Pall Mall smoker keeps birds.
+The owner of the yellow house smokes Dunhills.
+The man in the center house drinks milk.
+The Norwegian lives in the first house.
+The Blend smoker has a neighbor who keeps cats.
+The man who smokes Blue Masters drinks bier.
+The man who keeps horses lives next to the Dunhill smoker.
+The German smokes Prince.
+The Norwegian lives next to the blue house.
+The Blend smoker has a neighbor who drinks water.
+
+The question to be answered is: Who keeps fish?
+Answer adapted from https://gist.github.com/jgilchrist/1b85b04f4395057972dd
+*/
+
+// takes almost a minute (52s) to run atm on my machine with MemberO
+// takes about 5s to run with MemberO unrolled, _without any concurrency support_
+// the equivalent prolog program runs instantaneously
+func TestEinstein(t *testing.T) {
+	// not implementing counting right now, so numbers are symbols :)
+	one := ast.NewSymbol("one")
+	two := ast.NewSymbol("two")
+	three := ast.NewSymbol("three")
+	four := ast.NewSymbol("four")
+	five := ast.NewSymbol("five")
+
+	rightOf := func(x, y *ast.SExpr) micro.Goal {
+		return DisjPlus(
+			micro.ConjunctionO(
+				micro.EqualO(y, one),
+				micro.EqualO(x, two),
+			),
+			micro.ConjunctionO(
+				micro.EqualO(y, two),
+				micro.EqualO(x, three),
+			),
+			micro.ConjunctionO(
+				micro.EqualO(y, three),
+				micro.EqualO(x, four),
+			),
+			micro.ConjunctionO(
+				micro.EqualO(y, four),
+				micro.EqualO(x, five),
+			),
+		)
+	}
+
+	leftOf := func(x, y *ast.SExpr) micro.Goal {
+		return rightOf(y, x)
+	}
+
+	nextTo := func(x, y *ast.SExpr) micro.Goal {
+		return micro.DisjointO(
+			leftOf(x, y),
+			rightOf(x, y),
+		)
+	}
+
+	// NOTE: these should be anonymous vars, but we cannot make those atm without using call/fresh
+	// I'm creating vars with unique random names (from unix time) instead, for now
+	// somehow ast.NewVariable can introduce vars with scope: if you change one of the var names in
+	// street to duplicate another, you will see them substituted in variable bindings in state
+	// I don't think miniKanren _should_ work like that..
+	a := func() *ast.SExpr {
+		return ast.NewVariable(time.Now().String())
```

Comment:
There is a single non pure functional part of mini kanren, which is using ref in scheme.
This makes it possible to create unique variables, if I recall correctly
It might have something to do with what you do here.